sphericart/src/sphericart_cuda.cu

```diff
@@ -0,0 +1,149 @@
+
+    if (result) {
+        this->l_max_ = l_max;
+        this->grid_dim_x_ = GRID_DIM_X;
+        this->grid_dim_y_ = GRID_DIM_Y;
+        this->requires_grad_ = gradients;
+        this->requires_hessian_ = hessian;
+        this->scalar_size_ = scalar_size;
+    }
+    return result;
+}
+
+template <typename T>
+SphericalHarmonics<T>::SphericalHarmonics(size_t l_max, bool normalized) {
+    /*
+        This is the constructor of the SphericalHarmonics class. It initizlizes
+       buffer space, compute prefactors, and sets the function pointers that are
+       used for the actual calls
+    */
+
+    this->l_max = (int)l_max;
+    this->nprefactors = (int)(l_max + 1) * (l_max + 2);
+    this->normalized = normalized;
+    this->prefactors_cpu = new T[this->nprefactors];
+
+    // compute prefactors on host first
+    compute_sph_prefactors<T>((int)l_max, this->prefactors_cpu);
+    // allocate them on device and copy to device
+    CUDA_CHECK(
+        cudaMalloc(&this->prefactors_cuda, this->nprefactors * sizeof(T)));
+
+    CUDA_CHECK(cudaMemcpy(this->prefactors_cuda, this->prefactors_cpu,
+                          this->nprefactors * sizeof(T),
+                          cudaMemcpyHostToDevice));
+}
+
+template <typename T> SphericalHarmonics<T>::~SphericalHarmonics() {
+    // Destructor, frees the prefactors
+    delete[] (this->prefactors_cpu);
+    CUDA_CHECK(cudaFree(this->prefactors_cuda));
+}
+template <typename T>
+void SphericalHarmonics<T>::compute(const T *xyz, const size_t nsamples,
+                                    bool compute_with_gradients,
+                                    bool compute_with_hessian, T *sph, T *dsph,
+                                    T *ddsph) {
+
+    if (sph == nullptr) {
+        throw std::runtime_error(
+            "sphericart::cuda::SphericalHarmonics::compute expected "
+            "sph ptr initialised, instead nullptr found. Initialise "
+            "sph with cudaMalloc.");
+    }
+
+    if (compute_with_gradients && dsph == nullptr) {
+        throw std::runtime_error(
+            "sphericart::cuda::SphericalHarmonics::compute expected "
+            "dsph != nullptr since compute_with_gradients = true. "
+            "initialise dsph with cudaMalloc.");
+    }
+
+    if (compute_with_hessian && ddsph == nullptr) {
+        throw std::runtime_error(
+            "sphericart::cuda::SphericalHarmonics::compute expected "
+            "ddsph != nullptr since compute_with_hessian = true. "
+            "initialise ddsph with cudaMalloc.");
+    }
+
+    const std::lock_guard<std::mutex> guard(this->cuda_shmem_mutex_);
+
+    bool shm_result = this->cuda_shmem_.update_if_required(
+        sizeof(T), this->l_max, this->CUDA_GRID_DIM_X_, this->CUDA_GRID_DIM_Y_,
+        compute_with_gradients, compute_with_hessian);
+
+    if (!shm_result) {
+        printf("Warning: Failed to update shared memory specification with");
```

Comment:
This should use `std::cerr` to print to the standard error (or not print at all unless some environment variable is set to prevent standard stream pollution)